File: packages/listener-go/util/tx.go
package util

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/ethereum/go-ethereum/accounts/abi"
	"github.com/ethereum/go-ethereum/ethclient"
)

func CreateInterface(chain Chain) *ethclient.Client {
	client, err := ethclient.Dial(chain.RPCUrl)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to connect to RPC: %v", err)
		return nil
	}
	return client
}

func LoadSRC20Address() string {
	jsonPath := filepath.Join("..", "contracts", "out", "deploy.json")

	data, err := os.ReadFile(jsonPath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to read deploy.json: %v\n", err)
		return ""
	}

	var deployOut map[string]string
	if err := json.Unmarshal(data, &deployOut); err != nil {
		fmt.Fprintf(os.Stderr, "Failed to parse deploy.json: %v\n", err)
		return ""
	}

	return deployOut["MockSRC20"]
}

func ParseABI() *abi.ABI {
	contractAbi, err := abi.JSON(strings.NewReader(SRC20Abi))
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to parse ABI: %v\n", err)
		return nil
	}
	return &contractAbi
}

File: packages/listener-ts/src/util/tx.ts
import { http, type Chain } from 'viem';
import type { Account } from 'viem/accounts';
import { createShieldedWalletClient, getShieldedContract } from 'seismic-viem';

import { SRC20Abi } from './abi';
import DeployOut from '../../../contracts/out/deploy.json';

export async function createInterface(chain: Chain, account: Account) {
  const client = await createShieldedWalletClient({
    chain,
    account,
    transport: http(),
  });
  const contract = getShieldedContract({
    abi: SRC20Abi,
    address: DeployOut.MockSRC20 as `0x${string}`,
    client,
  });
  return { client, contract };
}
